File: backend/registrations/updateRegistration.mjs
import { GetCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { db, REGISTRATIONS_TABLE } from "./shared/db.mjs";
import { ok, badRequest, notFound, serverError, cors } from "./shared/response.mjs";
import { audit, callerFromEvent } from "./shared/auth.mjs";

export async function handler(event) {
  if (event.httpMethod === "OPTIONS") return cors();

  const caller = callerFromEvent(event);
  const { id } = event.pathParameters ?? {};
  if (!id) return badRequest("id is required");

  let body;
  try {
    body = JSON.parse(event.body ?? "{}");
  } catch {
    return badRequest("Invalid JSON");
  }

  try {
    const existing = await db.send(new GetCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
    }));
    if (!existing.Item) return notFound("Registration not found");

    const allowed = ["fullName", "organisation", "email", "phone", "position"];
    const updates = { updatedAt: new Date().toISOString() };
    allowed.forEach((k) => { if (k in body) updates[k] = body[k]; });

    const setExpr = Object.keys(updates).map((k) => `#${k} = :${k}`).join(", ");
    const exprNames = Object.fromEntries(Object.keys(updates).map((k) => [`#${k}`, k]));
    const exprValues = Object.fromEntries(Object.keys(updates).map((k) => [`:${k}`, updates[k]]));

    await db.send(new UpdateCommand({
      TableName: REGISTRATIONS_TABLE,
      Key: { registrationId: id },
      UpdateExpression: `SET ${setExpr}`,
      ExpressionAttributeNames: exprNames,
      ExpressionAttributeValues: exprValues,
      ReturnValues: "ALL_NEW",
    }));
    await audit("participant.edited", {
      entity: "participant", entityId: id,
      actorId: caller.actorId, actorLabel: caller.actorLabel,
    });
    return ok({ registrationId: id, ...updates });
  } catch (e) {
    console.error(e);
    return serverError();
  }
}

File: backend/registrations/updateRegistration.py
import json
import logging

from shared.auth import audit, caller_from_event, iso_now
from shared.db import registrations_table
from shared.response import bad_request, cors, not_found, ok, server_error, with_cors

log = logging.getLogger()

ALLOWED = ["fullName", "organisation", "email", "phone", "position"]


@with_cors
def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return cors()

    caller = caller_from_event(event)
    reg_id = (event.get("pathParameters") or {}).get("id")
    if not reg_id:
        return bad_request("id is required")

    try:
        body = json.loads(event.get("body") or "{}")
    except (ValueError, TypeError):
        return bad_request("Invalid JSON")

    try:
        existing = registrations_table().get_item(Key={"registrationId": reg_id}).get("Item")
        if not existing:
            return not_found("Registration not found")

        updates = {k: body[k] for k in ALLOWED if k in body}
        updates["updatedAt"] = iso_now()

        expr_names = {f"#{k}": k for k in updates}
        expr_values = {f":{k}": v for k, v in updates.items()}
        set_expr = ", ".join(f"#{k} = :{k}" for k in updates)

        registrations_table().update_item(
            Key={"registrationId": reg_id},
            UpdateExpression=f"SET {set_expr}",
            ExpressionAttributeNames=expr_names,
            ExpressionAttributeValues=expr_values,
            ReturnValues="ALL_NEW",
        )
        audit("participant.edited", entity="participant", entity_id=reg_id,
              actor_id=caller["actorId"], actor_label=caller["actorLabel"])
        return ok({**updates, "registrationId": reg_id})
    except Exception:
        log.exception("updateRegistration failed")
        return server_error()
